src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -0,0 +1,777 @@
+  nearFar: readonly number[];
+  sampleCount: number;
+  clipSpacePoints: readonly number[][];
+  interpolateFn: (fragData: FragData) => number[];
+}) {
+  const expected = new Float32Array(width * height * sampleCount * 4);
+
+  const viewport = [0, 0, width, height, ...nearFar];
+
+  const ndcPoints = clipSpacePoints.map(clipSpaceToNDC);
+  const windowPoints = ndcPoints.map(p => ndcToWindow(p, viewport));
+  const windowPoints2D = windowPoints.map(p => p.slice(0, 2));
+
+  const multisampleTable = kMultisamplingTables.get(sampleCount)!;
+  for (let y = 0; y < height; ++y) {
+    for (let x = 0; x < width; ++x) {
+      for (let s = 0; s < sampleCount; ++s) {
+        const fragmentPoint = [x + 0.5, y + 0.5];
+        const multisampleOffset = multisampleTable[s];
+        const sampleFragmentPoint = [x + multisampleOffset[0], y + multisampleOffset[1]];
+        const fragmentBarycentricCoords = calcBarycentricCoordinates(windowPoints2D, fragmentPoint);
+        const sampleBarycentricCoords = calcBarycentricCoordinates(
+          windowPoints2D,
+          sampleFragmentPoint
+        );
+
+        const output = interpolateFn({
+          fragmentPoint,
+          fragmentBarycentricCoords,
+          sampleBarycentricCoords,
+          clipSpacePoints,
+          ndcPoints,
+          windowPoints,
+        });
+
+        const offset = ((y * width + x) * sampleCount + s) * 4;
+        expected.set(output, offset);
+      }
+    }
+  }
+  return expected;
+}
+
+/**
+ * Computes 'builtin(position)`
+ */
+function computeFragmentPosition({
+  fragmentPoint,
+  fragmentBarycentricCoords,
+  clipSpacePoints,
+  windowPoints,
+}: FragData) {
+  return [
+    fragmentPoint[0],
+    fragmentPoint[1],
+    linearInterpolation(fragmentBarycentricCoords, getColumn(windowPoints, kZ)),
+    1 /
+      perspectiveInterpolation(
+        fragmentBarycentricCoords,
+        clipSpacePoints,
+        getColumn(clipSpacePoints, kW)
+      ),
+  ];
+}
+
+const kMultisamplingTables = new Map<number, readonly number[][]>([
+  [1, [[8 / 16, 8 / 16]]],
+  [
+    2,
+    [
+      [4 / 16, 4 / 16],
+      [12 / 16, 12 / 16],
+    ],
+  ],
+  [
+    4,
+    [
+      [6 / 16, 2 / 16],
+      [14 / 16, 6 / 16],
+      [2 / 16, 10 / 16],
+      [10 / 16, 14 / 16],
+    ],
+  ],
+  [
+    8,
+    [
+      [9 / 16, 5 / 16],
+      [7 / 16, 11 / 16],
+      [13 / 16, 9 / 16],
+      [5 / 16, 3 / 16],
+      [3 / 16, 13 / 16],
+      [1 / 16, 7 / 16],
+      [11 / 16, 15 / 16],
+      [15 / 16, 1 / 16],
+    ],
+  ],
+  [
+    16,
+    [
+      [9 / 16, 9 / 16],
+      [7 / 16, 5 / 16],
+      [5 / 16, 10 / 16],
+      [12 / 16, 7 / 16],
+
+      [3 / 16, 6 / 16],
+      [10 / 16, 13 / 16],
+      [13 / 16, 11 / 16],
+      [11 / 16, 3 / 16],
+
+      [6 / 16, 14 / 16],
+      [8 / 16, 1 / 16],
+      [4 / 16, 2 / 16],
+      [2 / 16, 12 / 16],
+
+      [0 / 16, 8 / 16],
+      [15 / 16, 4 / 16],
+      [14 / 16, 15 / 16],
+      [1 / 16, 0 / 16],
+    ],
+  ],
+]);
```

Comment:
I know that Shrek had a very similar table here https://github.com/gpuweb/cts/blob/main/src/webgpu/api/operation/render_pipeline/sample_mask.spec.ts#L143

The link that Shrek included in that instance is super helpful if you wanted to include it here also. (https://learn.microsoft.com/en-us/windows/win32/api/d3d11/ne-d3d11-d3d11_standard_multisample_quality_levels)

Also wondering if we should just move this to a more centralized place and provide a getter like: `getMultisampleTable(count: number)`. Would also be easier to read if the 16 was factored out and just divided in the end somehow.

Reply:
I got rid of the `/ 16` but I'm not convinced it's more readable. As it is, fractions seem clear to me (how far into the pixel the sample is) where-as integers feel like something I have to go read up on on what they are.  By a maybe poor analogy `1 / 64` (read as "one 64th") is much easier to understand than "0.015625" (what even is that?)

I guess I feel leaving them as x / 16 with a comment that these are fractional positions in a pixel might be more readable but for now I updated it.